File: swift/core/Sources/A2UICore/Models/SurfaceGroupModel.swift
// Copyright 2024 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import Combine
import Foundation
import OrderedCollections
import OrderedJSON

/// The root state model managing the collection of active surfaces.
///
/// `SurfaceGroupModel` owns the surface dictionary and surface lifecycle
/// (add/remove). It mirrors the `SurfaceGroupModel` type in the `web_core`
/// reference implementation.
@MainActor
public final class SurfaceGroupModel: ObservableObject {
  /// The map of active surfaces, published to the UI.
  @Published public private(set) var surfacesMap: [String: SurfaceViewModel] = [:]

  public init() {}

  // MARK: - Surface Lifecycle

  /// Adds a surface to the group.
  ///
  /// If a surface with the same ID already exists, the call is
  /// silently ignored (matching `web_core`'s behavior).
  public func addSurface(_ vm: SurfaceViewModel) {
    guard surfacesMap[vm.surfaceID] == nil else { return }
    surfacesMap[vm.surfaceID] = vm
  }

  /// Removes a surface from the group by its ID.
  public func removeSurface(id: String) {
    guard surfacesMap[id] != nil else { return }
    surfacesMap.removeValue(forKey: id)
  }

  // MARK: - Surface Lookup

  /// Retrieves a surface by its ID.
  public func surface(id: String) -> SurfaceViewModel? {
    surfacesMap[id]
  }

  /// Returns a snapshot of all active surfaces.
  public func allSurfaces() -> [String: SurfaceViewModel] {
    surfacesMap
  }

}

File: agent_sdks/python/a2ui_core/src/a2ui/core/state/surface_group_model.py
# Copyright 2024 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from typing import Dict, Optional
from ..common.events import EventSource, Subscription
from .surface_model import SurfaceModel


class SurfaceGroupModel:
    """The global manager and lifecycle container for all surfaces."""

    def __init__(self) -> None:
        self.surfaces: Dict[str, SurfaceModel] = {}
        self._surface_unsubscribers: Dict[str, Subscription] = {}

        self.on_surface_created = EventSource()
        self.on_surface_deleted = EventSource()
        self.on_action = EventSource()

    def get_surface(self, surface_id: str) -> Optional[SurfaceModel]:
        return self.surfaces.get(surface_id)

    def add_surface(self, surface: SurfaceModel) -> None:
        if surface.id in self.surfaces:
            return

        self.surfaces[surface.id] = surface
        # Forward actions from surface to global group listener and store subscription
        sub = surface.on_action.subscribe(lambda act: self.on_action.emit(act))
        self._surface_unsubscribers[surface.id] = sub

        self.on_surface_created.emit(surface)

    def delete_surface(self, surface_id: str) -> None:
        if surface_id in self.surfaces:
            surface = self.surfaces[surface_id]
            sub = self._surface_unsubscribers.pop(surface_id, None)
            if sub and hasattr(sub, "unsubscribe"):
                sub.unsubscribe()

            del self.surfaces[surface_id]
            surface.dispose()
            self.on_surface_deleted.emit(surface_id)

    @property
    def surfaces_map(self) -> Dict[str, SurfaceModel]:
        """Returns the dictionary of all active surfaces."""
        return self.surfaces

    def dispose(self) -> None:
        """Disposes of the group and all its surfaces."""
        for surface_id in list(self.surfaces.keys()):
            self.delete_surface(surface_id)
        if hasattr(self.on_surface_created, "dispose"):
            self.on_surface_created.dispose()
        if hasattr(self.on_surface_deleted, "dispose"):
            self.on_surface_deleted.dispose()
        if hasattr(self.on_action, "dispose"):
            self.on_action.dispose()
